features/48-user-profile-discovery/user-profile.md

```diff
@@ -0,0 +1,258 @@
+- `last_seen_at`: the timestamp when user last perform an action
+
+## New User Discovery API
+
+After revamping the user record and authentication object, a new user discovery
+API is proposed as a helper method query user records.
+
+The API signature will be `container.discoverUser(discoverData)` while
+`discoverData` is an object where each key-value pair will be transformed to
+an equality query predicate. The following snippet shows how it is transformed:
+
+```js
+const skygear = require('skygear');
+
+// the new User Discovery API
+skygear.discoverUser({ gender: 'male' })
+.then((users) => {
+  console.log(`${users.length} users founded`);
+});
+
+// the above equals to:
+const User = skygear.Record.extend('user');
+const q = skygear.Query(User);
+q.equalTo('gender', 'male');
+skygear.publicDB.query(q)
+.then((users) => {
+  console.log(`${users.length} users founded`);
+});
+```
+
+Also, the following user discovery methods are proposed as they are commonly
+used:
+
+- `container.discoverUserByUsername(username)`
+- `container.discoverUserByUsernames(usernames)`
+- `container.discoverUserByEmail(email)`
+- `container.discoverUserByEmails(emails)`
+
+## New Sign Up and Log In API
+
+### Sign Up API
+
+The new sign up API is proposed to enable: 1) sign up with custom data;
+2) update user profile during sign up. The new sign up API will perform:
+
+1. Discover user with the authentication data
+1. Sign up with the authentication data and password, if no user is found
+1. Update user record as profile
+
+The function signature will be `container.signup(authData, password, profile)`
+while `authData` is used to discover user. The following sign up API will be
+kept as helper methods:
+
+- `container.signupWithUsername(username, password)`
+- `container.signupWithEmail(email, password)`
+- `container.signupAnonymously()`
+
+### Log In API
+
+The similar signature is also proposed to log in API to enable log in with
+custom data: `container.login(authData, password)`.  The new log in API will
+perform:
+
+1. Discover user with the authentication data
+1. Log in with the authentication data and password, if only one user is found
+
+The current log in API will be kept as helper methods:
+
+- `container.loginWithUsername(username, password)`
+- `container.loginWithEmail(email, password)`
+
+### Authentication Record Configurations
+
+Since Skygear Server does not have a way for developers to set the uniqueness
+constraint of record fields (except writing the `before_save` hook of the
+record), configurations are proposed for authentication record.
+
+Authentication Record Configurations should contains: 1) Authentication Record
+Type; 2) Authentication Record Fields. They are suggested to be provided by
+environment variables during implementation:
+
+- `AUTH_RECORD_TYPE`: the record type for authentication, default: `user`
+- `AUTH_RECORD_FIELDS`: the record field names of the record for
```

Comment:
Expected that most of the user won't set `AUTH_RECORD_TYPE`. Since the current implementation allows changing the record type associating with the `_user` (default is `user` record), this env. var. just exposes it to developers, providing flexibility.

For `AUTH_RECORD_FIELDS`, any good suggestions ?